Script for running benchmarks on Java projects

romainbrenguier opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [na] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Add a script to run Java benchmarks

romainbrenguier committed:
Default JBMC arguments for benchmarks

Add a json file with some values for JBMC arguments.

romainbrenguier committed:
Gnuplot script for performance comparison

This can be used to compare performances of 2 runs of the benchmarks.
The data should be in 2 csv files named benchmark1.csv and
benchmark2.csv.

romainbrenguier committed:
Script to convert json benchmarks to csv

Add a script which takes the output of the benchmark_java_project.js
script (in json) and format it as a csv file.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b69f6e9f716f7a019d9d423031db862c1e93f99e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94334881

smowton commented:
Could you include a usage example?

peterschrammel reviewed (commented):
The scripts directory is getting a bit cluttered. Consider putting it into a sub-directory.

peterschrammel commented on scripts/benchmark_java_arguments.json:
```diff
@@ -0,0 +1,19 @@
+    "trace": true,
+    "cover": "location",
+    "classpath": ".",
+    "cover-function-only": true,
```
Looking at these parameters, I don't think these scripts are well placed in this repository.

allredj commented on scripts/benchmark_java_project.js:
```diff
@@ -0,0 +1,150 @@
+        + commandLine; 
+        const output =
+            execSync(timeCommand, { timeout: 1000 * config.timeout }).toString();
+        let execTime = fs.readFileSync('tmp_time.out', 'utf8').split('\n')[0];
```
The unit here seems to be seconds, whereas execTime below (:tiger2:) is in milliseconds. Should be made consistent.

allredj commented on scripts/benchmark_java_project.js:
```diff
@@ -0,0 +1,150 @@
+    } catch (err) {
+        return JSON.stringify({
+            classFile: classFile, function: functionName,
+            execTime: new Date() - startTime, success: false,
```
:tiger2:

tautschnig commented:
Is there any chance this could be re-implemented in Python? Given all the ongoing work around goto-symex I'd really like to use this tool, but 1) I don't know how to use it and 2) debugging/maintaining it might be even harder. Yet it seems to me that implementing this in Python would 1) not require any external tools (matplotlib or numpy would do the graphing work) and 2) possibly require less code.

But all this is wishlist stuff, the only really necessary steps are 1) a separate sub-directory and 2) a concrete usage example as others have said.

romainbrenguier committed:
README file on how to run benchmarks

This gives instruction on the installation and usage and an example of
command line to use.

romainbrenguier commented:
I moved the scripts to a new subdirectory and added a README

tautschnig reviewed (approved):
Seems ok to me.

tautschnig commented on scripts/benchmark/README.md:
```diff
@@ -0,0 +1,33 @@
+
+# Install
+
+    npm install
```
Actually that doesn't work for me, it complains about a missing `package.json` file.

romainbrenguier replied:
Sorry I forgot to add the file, should now be ok.

tautschnig commented on scripts/benchmark/README.md:
```diff
@@ -0,0 +1,33 @@
+
+From `apache-tika/tika-core/target/classes` run:
+
+    ./benchmark_java_project.js /path/to/jbmc -l method_list_example.txt -a java_arguments.json -m path/to/java-models-library/target/core-models.jar | tee result.json
```
Since you are including `java_arguments.json`, would you mind adding `method_list_example.txt`?

tautschnig commented on scripts/benchmark/README.md:
```diff
@@ -0,0 +1,33 @@
+
+# Example
+
+From `apache-tika/tika-core/target/classes` run:
```
It may be good to add full details on that.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: fd70d0a1bdebf2cfc0d7c509763c4cf019c3d448).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99410725
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

tautschnig commented:
Which version of nodejs is required to run this? I'm getting (on Ubuntu 14.04):
```
let program = require('commander');
    ^^^^^^^
SyntaxError: Unexpected identifier
    at Module._compile (module.js:439:25)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:902:3
```
with
```
$ nodejs --version
v0.10.25
```

romainbrenguier committed:
Add package.json

This is to be able to install dependencies easily, using npm.

romainbrenguier committed:
Add an example of method list for benchmarks

This are taken from the apache tika project. This can be used when the
apache tika package is compiled locally, otherwise this still serves as
an example of the syntax to use to provide a list of methods to the
script.

romainbrenguier commented:
> $ nodejs --version
> v0.10.25

Yes that's a bit old. Mine is v10.15.1. I've added that to the package.json

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 734a3aa0d606e5570e13a7b04d5dce1ddc05e761).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99424002
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 58bb18b3b5d4073cc4863ba81ac1e9d08ed59766).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99425354
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 6f2506fadaf8b1daef787c7d1d9a48252f8755a5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99431634

tautschnig reviewed (approved):
So I need to figure out a way to run this on one of my development boxes, but that shouldn't stand in the way of this being accepted.

peterschrammel commented on scripts/benchmark/benchmark_java_project.js:
```diff
@@ -0,0 +1,135 @@
+        + "It outputs information in the Json format.")
+    .option('-m, --models <model-path>', 'Path to the models library')
+    .option('-a, --arguments <json file>', 'Path to a json file describing arguments to give the executable')
+    .option('-l, --method-list <text file>', 'Text file from which to read the methods to test')
```
...methods to analyze?

peterschrammel commented on scripts/benchmark/benchmark_java_project.js:
```diff
@@ -0,0 +1,135 @@
+    .option('-m, --models <model-path>', 'Path to the models library')
+    .option('-a, --arguments <json file>', 'Path to a json file describing arguments to give the executable')
+    .option('-l, --method-list <text file>', 'Text file from which to read the methods to test')
+    .option('-s, --stop-after-symex', 'Only perform symex and do not generate tests')
```
...do not run the solver?

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 35350e059bb7b8f46858462f22139e145cd73b1d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99661759
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9ca0dff0c160feab5562c1c8ada3e7e01fba4ac1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99896834
